File: internal/model/mysql_backup.go
package model

import (
	"time"

	"gorm.io/gorm"
)

// MysqlBackupMode 备份模式：mysqldump 逻辑备份；xtrabackup 物理备份（经 SSH 执行 xtrabackup）。
const (
	MysqlBackupModeMysqldump   = "mysqldump"
	MysqlBackupModeXtrabackup  = "xtrabackup"
	MysqlBackupModeRemoteCheck = "remote_check" // 兼容旧数据，等同 xtrabackup
	// MysqlBackupExecXtrabackup 任务记录的实际执行方式
	MysqlBackupExecXtrabackup = "xtrabackup"
)

// MysqlBackupScope mysqldump 备份范围。
const (
	MysqlBackupScopeAll      = "all"
	MysqlBackupScopeDatabase = "database"
	MysqlBackupScopeTable    = "table"
)

// MysqlBackupTrigger 任务触发方式。
const (
	MysqlBackupTriggerManual    = "manual"
	MysqlBackupTriggerScheduled = "scheduled"
)

// MysqlBackupInstance 项目内 MySQL 备份目标（复用服务器 SSH 凭据）。
type MysqlBackupInstance struct {
	ID        uint   `json:"id" gorm:"primaryKey"`
	ProjectID uint   `json:"project_id" gorm:"not null;index:idx_mysql_bak_proj"`
	ServerID  uint   `json:"server_id" gorm:"not null;index"`
	Name      string `json:"name" gorm:"size:128;not null"`
	Enabled   bool   `json:"enabled" gorm:"default:true;index"`

	MysqlHost     string `json:"mysql_host" gorm:"size:255;not null;default:'127.0.0.1'"`
	MysqlPort     int    `json:"mysql_port" gorm:"not null;default:3306"`
	MysqlUser     string `json:"mysql_user" gorm:"size:128;not null"`
	EncPassword   string `json:"-" gorm:"type:longtext;comment:加密后的 MySQL 密码"`
	BackupMode    string `json:"backup_mode" gorm:"size:32;not null;default:'mysqldump'"`

	// BackupScope：all | database | table（mysqldump 模式）
	BackupScope string `json:"backup_scope" gorm:"size:32;not null;default:'all'"`
	// DatabaseName / BackupTable 单库、单表备份
	DatabaseName string `json:"database_name" gorm:"size:128"`
	BackupTable  string `json:"table_name" gorm:"column:table_name;size:128"`
	// DatabaseNames 历史字段：逗号分隔多库（未指定 scope 时兼容）
	DatabaseNames string `json:"database_names" gorm:"type:text"`

	RemoteDataDir string `json:"remote_data_dir" gorm:"size:512"`
	RemoteLogDir  string `json:"remote_log_dir" gorm:"size:512"`
	// MysqlDataDir 宿主机上 MySQL 真实 datadir（Docker 部署时填挂载目录，如 /export/mysql_data，非容器内 /var/lib/mysql）
	MysqlDataDir string `json:"mysql_datadir" gorm:"size:512"`
	// UploadToMinio 为 true 时将备份产物上传 MinIO；为 false 时仅执行备份/检查并保留远端文件。
	UploadToMinio bool `json:"upload_to_minio" gorm:"not null;default:true"`
	// MysqldumpWorkDir 远端 mysqldump 落盘目录（绝对路径，非 /tmp）
	MysqldumpWorkDir string `json:"mysqldump_work_dir" gorm:"size:512"`
	// MysqldumpOptions JSON 字符串数组，选项 id 见 mysqlbackup.MysqldumpOptionCatalog
	MysqldumpOptions string `json:"mysqldump_options" gorm:"type:text"`
	// MysqldumpExtraArgs 额外 mysqldump 参数（空格分隔，须以 - 开头）
	MysqldumpExtraArgs string `json:"mysqldump_extra_args" gorm:"size:512"`

	ScheduleEnabled bool       `json:"schedule_enabled" gorm:"not null;default:false;index"`
	CronSpec        string     `json:"cron_spec" gorm:"size:256;not null;default:''"`
	LastScheduledAt *time.Time `json:"last_scheduled_at,omitempty"`

	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `json:"-" gorm:"index"`
}

func (MysqlBackupInstance) TableName() string { return "mysql_backup_instances" }

// MysqlBackupJob 单次备份任务记录。
type MysqlBackupJob struct {
	ID         uint   `json:"id" gorm:"primaryKey"`
	InstanceID uint   `json:"instance_id" gorm:"not null;index"`
	ProjectID  uint   `json:"project_id" gorm:"not null;index"`
	Status     string `json:"status" gorm:"size:32;not null;index"` // pending/running/success/failed
	BackupMode string `json:"backup_mode" gorm:"size:32"`
	TriggerType string `json:"trigger_type" gorm:"size:32;not null;default:'manual'"`
	BackupScope string `json:"backup_scope,omitempty" gorm:"size:32"`
	DatabaseName string `json:"database_name,omitempty" gorm:"size:128"`
	BackupTable  string `json:"table_name,omitempty" gorm:"column:table_name;size:128"`

	RemotePath   string `json:"remote_path,omitempty" gorm:"size:512"`
	MinioBucket  string `json:"minio_bucket,omitempty" gorm:"size:128"`
	MinioObject  string `json:"minio_object,omitempty" gorm:"size:512"`
	FileSize     int64  `json:"file_size" gorm:"default:0"`
	CheckOK      bool   `json:"check_ok" gorm:"default:false"`
	LogExcerpt   string `json:"log_excerpt,omitempty" gorm:"type:text"`
	ErrorMessage string `json:"error_message,omitempty" gorm:"type:text"`

	StartedAt  *time.Time `json:"started_at,omitempty"`
	FinishedAt *time.Time `json:"finished_at,omitempty"`
	CreatedAt  time.Time  `json:"created_at"`
}

func (MysqlBackupJob) TableName() string { return "mysql_backup_jobs" }

File: web/src/services/mysql-backup.ts
import { getData, http } from "./http";

export type MysqlBackupScope = "all" | "database" | "table";

export interface MysqlBackupInstance {
  id: number;
  project_id: number;
  server_id: number;
  server_name?: string;
  name: string;
  enabled: boolean;
  mysql_host: string;
  mysql_port: number;
  mysql_user: string;
  backup_mode: "mysqldump" | "remote_check" | string;
  backup_scope?: MysqlBackupScope | string;
  database_name?: string;
  table_name?: string;
  database_names?: string;
  remote_data_dir?: string;
  remote_log_dir?: string;
  mysql_datadir?: string;
  upload_to_minio?: boolean;
  mysqldump_work_dir?: string;
  mysqldump_options?: string[];
  mysqldump_extra_args?: string;
  schedule_enabled?: boolean;
  cron_spec?: string;
  last_scheduled_at?: string;
  created_at?: string;
  updated_at?: string;
}

export type MysqlBackupInstancePayload = {
  server_id: number;
  name: string;
  enabled?: boolean;
  mysql_host?: string;
  mysql_port?: number;
  mysql_user: string;
  mysql_password?: string;
  backup_mode?: string;
  backup_scope?: MysqlBackupScope | string;
  database_name?: string;
  table_name?: string;
  database_names?: string;
  remote_data_dir?: string;
  remote_log_dir?: string;
  mysql_datadir?: string;
  upload_to_minio?: boolean;
  mysqldump_work_dir?: string;
  mysqldump_options?: string[];
  mysqldump_extra_args?: string;
  schedule_enabled?: boolean;
  cron_spec?: string;
};

export type MysqldumpOptionItem = { id: string; label: string; flag: string; group?: string };

export function listMysqldumpOptions(projectId: number) {
  return getData<MysqldumpOptionItem[]>(http.get(`/projects/${projectId}/mysql-backup/mysqldump-options`));
}

export interface MysqlBackupJob {
  id: number;
  instance_id: number;
  project_id: number;
  status: string;
  backup_mode?: string;
  trigger_type?: string;
  backup_scope?: string;
  database_name?: string;
  table_name?: string;
  remote_path?: string;
  minio_bucket?: string;
  minio_object?: string;
  file_size?: number;
  check_ok?: boolean;
  log_excerpt?: string;
  error_message?: string;
  started_at?: string;
  finished_at?: string;
  created_at?: string;
}

export function listMysqlBackupInstances(projectId: number, params?: { page?: number; page_size?: number }) {
  return getData<{ list: MysqlBackupInstance[]; total: number }>(
    http.get(`/projects/${projectId}/mysql-backup/instances`, { params }),
  );
}

export function createMysqlBackupInstance(projectId: number, payload: MysqlBackupInstancePayload) {
  return getData<MysqlBackupInstance>(http.post(`/projects/${projectId}/mysql-backup/instances`, payload));
}

export function updateMysqlBackupInstance(projectId: number, instanceId: number, payload: MysqlBackupInstancePayload) {
  return getData<MysqlBackupInstance>(http.put(`/projects/${projectId}/mysql-backup/instances/${instanceId}`, payload));
}

export function deleteMysqlBackupInstance(projectId: number, instanceId: number) {
  return getData<{ deleted: boolean }>(http.delete(`/projects/${projectId}/mysql-backup/instances/${instanceId}`));
}

export function pingMysqlBackupInstance(projectId: number, instanceId: number) {
  return getData<{ ok: boolean; message: string }>(http.post(`/projects/${projectId}/mysql-backup/instances/${instanceId}/ping`));
}

export function checkMysqlRemoteBackup(projectId: number, instanceId: number) {
  return getData<{ ok: boolean; message: string; backup_file?: string; log_file?: string }>(
    http.post(`/projects/${projectId}/mysql-backup/instances/${instanceId}/check-remote`),
  );
}

export function runMysqlBackup(projectId: number, instanceId: number) {
  return getData<MysqlBackupJob>(http.post(`/projects/${projectId}/mysql-backup/instances/${instanceId}/run`));
}

export function listMysqlBackupJobs(projectId: number, params?: { instance_id?: number; page?: number; page_size?: number }) {
  return getData<{ list: MysqlBackupJob[]; total: number; page: number; page_size: number }>(
    http.get(`/projects/${projectId}/mysql-backup/jobs`, { params }),
  );
}

export function presignMysqlBackupJob(projectId: number, jobId: number) {
  return getData<{ url: string }>(http.get(`/projects/${projectId}/mysql-backup/jobs/${jobId}/presign`));
}
